sample.fvwmrc/BradyMontz/src/xpaste-1.1/xpaste.man
.TH xpaste 1 "May, 1990"
.SH xpaste
xpaste \- pop up a window within which the contents of the primary selection
buffer is pasted.
.SH SYNOPSIS
.B xpaste
.SH DESCRIPTION
Xpaste will create a window containing the current contents of the primary
selection buffer. The size of this window is dictated by the length of
the longest line in the selection, the number of lines in the selection,
and the size of the current font. Variable-width fonts are supported.
.PP
Clicking on the \fBDONE\fP button causes the window to go away.
.SH OPTIONS
Xpaste accepts all the standard X Toolkit command line options. 
.SH WIDGETS
The hierarchy of the widgets composing \fBxpaste\fP follow. The widget class
name is given first, followed by the widget instance name.
.sp
.nf
.ta .5i 1.0i 1.5i 2.0i
XPaste	xpaste
	Form	 form
		Command	 done
		AsciiString	 text
.SH SEE ALSO
X(1), xclipboard(1)
.SH HISTORY
Written by John DiMarco at the University of Toronto.
.SH BUGS
Selections containing tabs and/or binary data may cause
.B xpaste 
to size the window incorrectly. 


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

sample.fvwmrc/BradyMontz/src/xpaste-1.1/xpaste.c
/* 
 * Xpaste - grab the current primary selection, and pop up a window containing
 *          that selection. 
 *        - print button added by Stewart A. Levin 6/20/91
 *
 */

#ifndef lint
static char rcsid[] = "$Id$";
#endif

#include <stdio.h>
#include <X11/StringDefs.h>
#include <X11/Xlib.h>
#include <X11/Intrinsic.h>
#include <X11/Xatom.h>
#include <X11/Xaw/Text.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Command.h>
#include <varargs.h>
#include "patchlevel.h"

#ifndef WIDTHPADDING
#define WIDTHPADDING 6
#endif /* WIDTHPADDING */
#ifndef HEIGHTPADDING
#define HEIGHTPADDING 6
#endif /* HEIGHTPADDING */

void Xit();
void Printit();
void GetDimensions();
int GetPrimarySelection();

char *progname;
char **av;
int ac;

static XtCallbackRec done[] = {
	{Xit, NULL},
	{NULL,NULL},
};
static XtCallbackRec print[] = {
	{Printit, NULL},
	{NULL,NULL},
};

Widget CreateWidget();
void ConfigureWidget();

static char *buff;
static Widget root, text, fm, button[3];

static struct xpaste_res {
    XFontStruct *font;
    char *printcmd;
    } appres;

typedef struct xpaste_res *xpres;
extern char *getenv();

main(argc, argv)
int argc;
char *argv[];
{

	Dimension height, width;

	progname=argv[0];
	av=argv;
	ac=argc;
	root = XtInitialize(argv[0], "XPaste", NULL, 0, &argc, argv);

	(void)GetPrimarySelection(XtDisplay(root), &buff);

	/* 
	 * If there is no current selection, exit quietly. 
	 */
	if(NULL==buff) (void)exit(0);
	
	/*
	 * Compute the extents of the text. The eventual width of the text 
	 * widget will be the length (in pixels) of the longest line. Variable
	 * width fonts are handled correctly. The eventual height of the text
	 * widget will be the sum of the heights of the lines in the selection.
	 * Yes, computing the height for a single line and multiplying by the
	 * number of lines would be more efficient. But there's no server
	 * roundtrip required by XTextExtents, so the cost isn't high. And who
	 * knows: maybe we'll see variable height fonts?  
	 */

	{
		/* Get the default font */
		static XtResource resources[] = {
			{"font", XtCFont, XtRFontStruct, sizeof(XFontStruct *),
			XtOffset(xpres,font), XtRString, "fixed"},
			{"printcmd", XtCString, XtRString, sizeof(char *),
			XtOffset(xpres,printcmd), XtRString, "lpr"},
		};
		XtGetApplicationResources( root, (caddr_t)(&appres), 
				resources, XtNumber(resources), 
				(ArgList)NULL, (Cardinal)0      );
	}
	GetDimensions(buff, appres.font, &width, &height);

	/* 
	 * now create the text widget, using the computed width and height and 
	 * the selection string.
	 */
	fm = CreateWidget("form", formWidgetClass, root, 
					   XtNdefaultDistance,  0,
					   NULL);
	button[0] = CreateWidget("done", commandWidgetClass, fm,
					   XtNcallback,         done,
					   XtNresizable,        False,
					   NULL);
	button[1] = CreateWidget("print", commandWidgetClass,fm,
					   XtNcallback,         print,
					   XtNresizable,        False,
					   XtNfromHoriz,        button[0],
					   NULL);
	text = CreateWidget("text", asciiTextWidgetClass, fm,
					   XtNwidth,            width,
					   XtNheight,           height,
					   XtNstring,           buff,
					   XtNfont,             appres.font,
					   XtNfromVert,         button[0],
					   XtNborderWidth,      0,
					   XtNresizable,        True,
					   XtNdisplayCaret,	False,
					   NULL); 

	XtRealizeWidget(root);
	XtMainLoop();
}

void Xit(w, closure, call_data)
Widget w;
caddr_t closure;
caddr_t call_data;
{
	/* exit program */
	(void)exit(0);
}

void Printit(w, closure, call_data)
Widget w;
caddr_t closure;
caddr_t call_data;
{
	FILE *syscmd;
	if(NULL==(syscmd=popen(appres.printcmd,"w"))) {
		(void)XtWarning("Can't create pipe\n");
	} else {
		(void) fprintf(syscmd,"%s\n",buff);
		(void) pclose(syscmd);
	}
}

void GetDimensions(string, font, width, height)
char *string;
XFontStruct *font;
Dimension *width, *height;
{
	/* Compute the width and height of the given string when    */
	/* displayed using the specified font. It does the right    */
	/* thing for both variable width and height fonts, and      */
	/* multiple-line (newline) strings. TABS are not supported. */
	char *p, *b;
	int dir, ascent, descent;
	XCharStruct all;

	p=b=string;
	*width=*height=0;

	while(True){
		if(!*p || '\n'==*p){
			Dimension tmp;

			/* height computation */
			XTextExtents(font,b,p-b,&dir,&ascent,&descent,&all);
			*height+=(Dimension)(ascent+descent);

			/* width computation */
			tmp=(Dimension)XTextWidth(font, b, p-b);
			if(tmp>*width) *width=tmp;

			b=p+1;
			if(!*p) break;
		}
		p++;
	}
	*width+=WIDTHPADDING;
	*height+=HEIGHTPADDING;
}

int GetPrimarySelection(display, string)
Display *display;
char **string;
{
	/* 
	 * Get and return the primary selection. This clobbers cut buffer 0.
	 * The return value is the number of bytes.
	 */

	int bytes;
	Window root;

	/* 
	 * Need to convert PRIMARY selection to CUT_BUFFER0 because 
	 * XFetchBytes only knows how to get contents of cut buffer 0. I 
	 * could have used the Xt selection functions to grab the PRIMARY
	 * selection directly, but they're unnecessarily complicated.
	 * 
	 * Note: gnuemacs doesn't use the primary selection, so this won't
	 *       work in an emacs window. 
	 */
	
	root=XDefaultRootWindow(display);
	XConvertSelection(display, XA_PRIMARY, XA_CUT_BUFFER0, None, root,
			  CurrentTime);
	(*string)=XFetchBytes(display, &bytes);
	return(bytes);
}
